### PR TITLE
Value object's hashCode, equals, toString methods

### PR DESCRIPTION
**Pull request by hurelhuyag**

### CODE REVIEW

File: src/main/resources/codegen/java/ValueDataObject.ftl [hunk 1]
```diff
@@ -34,4 +38,37 @@ public class ${dataValueObjectName} {
...
+              </#list>
+              .toHashCode();
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (obj == this){
+      return true;
+    }
+    if (!(obj instanceof ${valueObjectName})){
```
Comment (VaughnVernon) [inline src/main/resources/codegen/java/ValueDataObject.ftl:55]: Must be:

```
if (getClass() != other.getClass()) {
  return false;
}
```

File: src/main/resources/codegen/java/ValueDataObject.ftl [hunk 2]
```diff
@@ -34,4 +38,37 @@ public class ${dataValueObjectName} {
...
+  public int hashCode() {
+    return new HashCodeBuilder()
+              <#list memberNames as member>
+              .append(${member})
+              </#list>
+              .toHashCode();
+  }
+
+  @Override
+  public boolean equals(Object obj) {
```
Comment (VaughnVernon) [inline src/main/resources/codegen/java/ValueDataObject.ftl:51]: Rename `obj` to `other`.

File: src/main/resources/codegen/java/ValueDataObject.ftl [hunk 3]
```diff
@@ -34,4 +38,37 @@ public class ${dataValueObjectName} {
...
+    if (obj == this){
+      return true;
+    }
+    if (!(obj instanceof ${valueObjectName})){
+      return false;
+    }
+    ${valueObjectName} another = (${valueObjectName}) obj;
+    return new EqualsBuilder()
```
Comment (VaughnVernon) [inline src/main/resources/codegen/java/ValueDataObject.ftl:59]: Seed with prime number `31`.

**Review by VaughnVernon — CHANGES_REQUESTED** [review]

File: src/main/resources/codegen/java/ValueDataObject.ftl (hunk 3 above)
Comment (hurelhuyag) [reply src/main/resources/codegen/java/ValueDataObject.ftl:59]: For equals() or hashCode()? Why 31? HashCodeBuilder itself uses another prime number internally.

Comment (VaughnVernon) [reply src/main/resources/codegen/java/ValueDataObject.ftl:59]: @hurelhuyag https://stackoverflow.com/a/299748

**Review by VaughnVernon — CHANGES_REQUESTED** [review]

File: src/main/resources/codegen/java/ValueDataObject.ftl (hunk 3 above)
Comment (hurelhuyag) [reply src/main/resources/codegen/java/ValueDataObject.ftl:59]: @VaughnVernon There is no difference between 37 and 31. Maybe 37 can be better than 31. But I changed it to 31.

**Comment by VaughnVernon** [comment]
Maybe you've heard of Josh Bloch. I'll take his advice on Java.

**Comment by VaughnVernon** [comment]
Hmmm, I just noticed that you forgot to put a space between the ending `)` and the opening `{` here `if (...) {`. Or is that your personal non-standard style? I've never seen that in 23 years of programming in Java, or 37 with any C-based syntaxes.

Please fix.

https://github.com/vlingo/xoom-designer/blob/dceef258de0326895f60de38fd10eea881d950e7/src/main/resources/codegen/java/ValueObject.ftl#L40

https://github.com/vlingo/xoom-designer/blob/dceef258de0326895f60de38fd10eea881d950e7/src/main/resources/codegen/java/ValueObject.ftl#L43

**Pull request merged**